### ISSUE TITLE
Accumulate autogenerated migrations over multiple calls to `run_migrations()`

### ISSUE THREAD

**Issue by StreakyCobra** [sqlalchemy/alembic#658]
I am creating a multi-tenant application, with several tenant-specific schema and also public shared tables. For this I made a setup that works quite well. I'm now only struggling with the auto-generation of migrations. 

I have simplified my use case to this minimal example. In `run_migrations_online`, I call several times the `context.run_migrations()`, every time with a different `target_metadata` that only concerns the table that are linked to one schema:
 
```python
def run_migrations_online():

    connectable = create_engine(os.environ["DATABASE_URL"])

    with connectable.connect() as connection:

        schema = Schema.PUBLIC

        print()
        print("-" * 80)
        print(f"Migrating schema {schema.schema_name}\n")

        context.configure(
            connection=connection,
            target_metadata=schema.base.metadata,
            include_schemas=True,
        )

        with context.begin_transaction():
            context.run_migrations()

        schema = Schema.TENANT

        print()
        print("-" * 80)
        print(f"Migrating schema {schema.schema_name}\n")

        context.configure(
            connection=connection,
            target_metadata=schema.base.metadata,
            include_schemas=True,
        )

        with context.begin_transaction():
            context.run_migrations()

```

When running `alembic revision --autogenerate -m "create tables"`, we can see that alembic detect correctly the modifications (one table added) for each call:

```
--------------------------------------------------------------------------------
Migrating schema public

INFO  [alembic.runtime.migration] Context impl PostgresqlImpl.
INFO  [alembic.runtime.migration] Will assume transactional DDL.
INFO  [alembic.autogenerate.compare] Detected added table 'exchange_rate'

--------------------------------------------------------------------------------
Migrating schema tenant

INFO  [alembic.runtime.migration] Context impl PostgresqlImpl.
INFO  [alembic.runtime.migration] Will assume transactional DDL.
INFO  [alembic.autogenerate.compare] Detected added table 'tenant.inventory'
```

But the problem is that when it generate the migration script, it keeps only the modifications detected during the last call of `context.run_migrations()`, namely `tenant.inventory` table in this case:

```python
def upgrade():
    # ### commands auto generated by Alembic - please adjust! ###
    op.create_table('inventory',
    sa.Column('id', sa.Integer(), nullable=False),
    sa.PrimaryKeyConstraint('id'),
    schema='tenant'
    )
    # ### end Alembic commands ###


def downgrade():
    # ### commands auto generated by Alembic - please adjust! ###
    op.drop_table('inventory', schema='tenant')
    # ### end Alembic commands ###
```

Now if I invert the order in which TENANT and PUBLIC are processed, both changes are correctly discovered, but in this case the migration script only keep tracks of `public.exchange_rate` table:

```
--------------------------------------------------------------------------------
Migrating schema tenant

INFO  [alembic.runtime.migration] Context impl PostgresqlImpl.
INFO  [alembic.runtime.migration] Will assume transactional DDL.
INFO  [alembic.autogenerate.compare] Detected added table 'tenant.inventory'

--------------------------------------------------------------------------------
Migrating schema public

INFO  [alembic.runtime.migration] Context impl PostgresqlImpl.
INFO  [alembic.runtime.migration] Will assume transactional DDL.
INFO  [alembic.autogenerate.compare] Detected added table 'exchange_rate'
```

```python
def upgrade():
    # ### commands auto generated by Alembic - please adjust! ###
    op.create_table('exchange_rate',
    sa.Column('date', sa.DateTime(), nullable=False),
    sa.PrimaryKeyConstraint('date'),
    schema='public'
    )
    # ### end Alembic commands ###


def downgrade():
    # ### commands auto generated by Alembic - please adjust! ###
    op.drop_table('exchange_rate', schema='public')
    # ### end Alembic commands ###
```

In other words the auto-generate features generate the script only for the last run of `context.run_migrations()`.

---

**Question**:

Is it somehow possible to accumulate the changes over multiple calls to `context.run_migrations()` when using `autogenerate` feature?

(«No» is a perfectly valid answer, at least I would know I should stop trying :smile:)

_Some notes_:

- I have to keep this structure where I reconfigure the context several time, because when I have multiple tenants I actually want one `version_table` per tenant, and this is set using `context.configure()`.
- Online upgrade (with `alembic upgrade head`) does not have this problem because the changes are applied directly, there is no need to accumulate them.

**Comment by zzzeek** [comment]
are you trying to generate a single migration script for the different schemas?    to support the use case of multiple metadata collections you can send a list of them to the target_metadata parameter:     https://alembic.sqlalchemy.org/en/latest/autogenerate.html#autogenerating-multiple-metadata-collections   but that implies they'd each have a "schema" qualifier set up appropriately.

from your example it looks like you are autogenerating and scanning all schemas each time you run, I would imagine this generates DROPs for all the elements that aren't in the local metadata so not sure how you are handling that part unless you have an include_object hook also.

since you're doing everything on one connection it looks like you should just pass a sequence of MetaData objects to target_metadata.

**Comment by StreakyCobra** [comment]
> are you trying to generate a single migration script for the different schemas?

Yes

> but that implies they'd each have a "schema" qualifier set up appropriately.

That's the case :)

> from your example it looks like you are autogenerating and scanning all schemas each time you run, I would imagine this generates DROPs for all the elements that aren't in the local metadata so not sure how you are handling that part unless you have an include_object hook also.

I have defined an `include_object` method as well yes :) But you made me realize that this is no longer needed. I just removed the `include_schemas` and `include_object` altogether and it works the same.

> since you're doing everything on one connection it looks like you should just pass a sequence of MetaData objects to target_metadata.

I can do this, and indeed it detects both changes:

```
INFO  [alembic.runtime.migration] Context impl PostgresqlImpl.
INFO  [alembic.runtime.migration] Will assume transactional DDL.
INFO  [alembic.autogenerate.compare] Detected added table 'exchange_rate'
INFO  [alembic.autogenerate.compare] Detected added table 'tenant.inventory'
```

And generates the migration script accordingly. So it's kinda what I want for the `autogenerate` feature :+1: 

---

**Follow-up question**:

Now for applying the migrations, I need a schema-specific alembic version table. For this I am doing two things.

First, in order to have the same migration script applying both on multiple tenants and on public, I edit the migration script to apply each part conditionally: 

```python
def upgrade():
    if os.environ.get("TENANT"):
        op.create_table(
            "inventory",
            sa.Column("date", sa.Integer(), nullable=False),
            sa.PrimaryKeyConstraint("id"),
            schema=os.environ["TENANT"],
        )
    elif os.environ.get("SCHEMA") == "public":
        op.create_table(
            "exchange_rate",
            sa.Column("date", sa.DateTime(), nullable=False),
            sa.PrimaryKeyConstraint("date"),
            schema="public",
        )
```

This is my responsibility to adapt it correctly and I don't expect alembic to detect it in any way.

Second my `run_migrations_online()` script run this way:

1. Run the migration of `public` schema
2. Run several time the tenant migration, once for each tenant

Note that this make use of a different `version_table` for each tenant, so different tenant can be in a different states, and if I create a new empty tenant schema and alembic will detect that this tenant is running behind and migrate it from beginning:

```python
def run_migrations_online():
    connectable = create_engine(os.environ["DATABASE_URL"])

    with connectable.connect() as connection:

        # FIRST MIGRATE PUBLIC

        schema = Schema.PUBLIC
        os.environ["SCHEMA"] = "public"
        os.environ["TENANT"] = ""

        print()
        print("-" * 80)
        print(f"Migrating public\n")

        context.configure(
            connection=connection,
            target_metadata=schema.base.metadata,
        )

        with context.begin_transaction():
            context.run_migrations()

        # THEN MIGRATE EACH TENANT INDIVIDUALLY WITH THEIR OWN VERSION TABLE

        for schema in Schema.tenants():

            os.environ["SCHEMA"] = ""
            os.environ["TENANT"] = schema.schema_name

            print()
            print("-" * 80)
            print(f"Migrating tenant {schema.schema_name}\n")

            context.configure(
                connection=connection,
                target_metadata= schema.base.metadata,
                version_table=f"version_{schema.schema_name}",
                version_table_schema="alembic",
            )

            with context.begin_transaction():
                context.run_migrations()
```

This approach is working quite well for me.

--- 

So it seems that I have one approach that runs well for auto-generating the migration scripts, and another method that work well for applying the migration. I can definitively keep two different functions for that, and it probably make sense actually, so **is there a way to to check in we are running in the `revision --autogenerate` or `upgrade` command to choose the function accordingly?**

I found a `hack`, but hopefully there is a better solution :)

```python
    if (
        hasattr(config.cmd_opts, "cmd")
        and len(config.cmd_opts.cmd) >= 3
        and "autogenerate" in config.cmd_opts.cmd[2]
    ):
       ...
```

**Comment by zzzeek** [comment]
cmd_opts is the options, I'd look for the command name "revision" rather than look for the "autogenerate" argument though.   

```
from alembic import command
if config.cmd_opts.cmd[0] is command.revision:
   # ...

```

**Comment by StreakyCobra** [comment]
Perfect, thanks a lot for taking the time to help me. I'll go forward with this :+1:
